include/rxwayland/signal.hpp

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <cassert>
+#include <memory>
+
+#include "rxcpp/rx.hpp"
+
+#include "wayland-server-core.h"
+
+namespace rxwayland {
+namespace detail {
+
+template <class T>
+struct listener_with_subscriber {
+  using this_t = listener_with_subscriber<T>;
+
+  wl_listener listener;
+  wl_listener destroy_listener;
+  rxcpp::subscriber<T> subscriber;
+
+  listener_with_subscriber(const rxcpp::subscriber<T>& s)
+      : listener{}, destroy_listener{}, subscriber{s} {
+    listener.notify = [](wl_listener* l, void* data) {
+      this_t* this_ptr = wl_container_of(l, this_ptr, listener);
+      this_ptr->subscriber.on_next(static_cast<T>(data));
+    };
+    destroy_listener.notify = [](wl_listener* l, void*) {
+      this_t* this_ptr = wl_container_of(l, this_ptr, destroy_listener);
+      this_ptr->subscriber.on_completed();
+    };
+  }
+};
+
+}  // namespace detail
+
+template <class T>
+rxcpp::observable<T> from_signal(wl_signal* signal, wl_signal* destroy_signal) {
```

Comment:
libweston(from wayland reference compositor, weston) or other third party libs support various signals.
Note that these signals only emitted on the thread which runs event loop.
Most functions and structures supported by libwestonor third party libs are NOT THREAD SAFE.